### ISSUE TITLE
Classification for non-square images doesn't work properly

### ISSUE THREAD

**Issue by agirbau** [AlexeyAB/darknet#2350]
Hello Alexey!
First of all, thanks for the amazing work! I've been trying lately the classification branch and I found something curious.

If I try to train a classifier with squared images (e.g. 64x64) it trains correctly.
![image](https://user-images.githubusercontent.com/23292221/52280401-1ce4d680-295c-11e9-9b1d-82a54ba5421c.png)

If, instead, I change the input size of the network to a non-squared image (e.g 96 x 64) it doesn't train at all (or at least it's not correctly printed).
![image](https://user-images.githubusercontent.com/23292221/52280523-66cdbc80-295c-11e9-8982-8a959b202294.png)

Do you know why this may happen? Thanks a lot!

**Comment by AlexeyAB** [comment]
@agirbau Hi,

* What command do you use for train Classifier?

* What dataset do you use?

* What cfg-file do you use?

**Comment by agirbau** [comment]
Hello @AlexeyAB!

> * What command do you use for training Classifier?

I work in Ubuntu 16.04:
./darknet classifier train cfg/classifier.data cfg/yolov3-classifier.cfg

> * What dataset do you use?

I use a custom dataset containing 2 person classes (class1 and class2)

> * What cfg-file do you use?

I use a custom cfg file with little amount of layers. 

**The network configuration is:**
[net]
batch=32
subdivisions=4
#height=48
#width=48
height=80
width=48							

channels=3
momentum=0.9
decay=0.0005
angle=10
saturation = 0
exposure = 1.5
hue=0
max_crop=24

learning_rate=0.001
max_batches = 5000
policy=steps
steps=600,2000,3500
scales=10,.1,.1

**The classification layer is:**
[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=64
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=2
activation=relu

[avgpool]

[softmax]
groups=1

[cost]
type=sse

Do you have a clue on what could be happening? Thanks a lot!

**Comment by agirbau** [comment]
Of course, in the top comment, I use the intermediate convolutional, max-pooling and stuff layers.

**Comment by AlexeyAB** [comment]
@agirbau 

Try to add this code: 
```cpp
    char buff[256];
    static int id = 0;
    sprintf(buff, "crop_%d", ++id);
    save_image(crop, buff);
```

here: https://github.com/AlexeyAB/darknet/blob/64b217aa868f6f050ecdb0b627aeda3dc866912d/src/image.c#L1599

compile, run training and look what is in the `crop_x.jpg` images when you use square and non-sqaure network.

----

Also you set:

> width=48
> max_crop=24

https://github.com/AlexeyAB/darknet/blob/64b217aa868f6f050ecdb0b627aeda3dc866912d/src/parser.c#L657

so `min_crop = 48 > max_crop = 24`

Try to comment `max_crop = 24` in your cfg.

**Comment by agirbau** [comment]
Hello @AlexeyAB !

I added the code and removed the max_crop=24 from the configuration file, the augmented data doesn't seem to have errors.

**Original image**
![img_train](https://user-images.githubusercontent.com/23292221/52356572-c4353c80-2a34-11e9-943b-a43ad05a07c8.jpg)

**Inputting a square image**
![img_crop_train](https://user-images.githubusercontent.com/23292221/52356571-c4353c80-2a34-11e9-8da9-b7917fab34fd.jpg)

**Inputting a non-square image**
![img_crop_nonsquare](https://user-images.githubusercontent.com/23292221/52356568-c4353c80-2a34-11e9-8998-7241aed5edd8.jpg)

**Comment by AlexeyAB** [comment]
@agirbau Hi,
Do you still get training issue without max_crop=24 ?
Can you rename your cfg-file to the txt file and drag-n-drop to your message?

**Comment by agirbau** [comment]
Hello @AlexeyAB. First of all, apologies for the belated response (I didn't had access to the PC this week).

### cfg file configuration ###

[net]
batch=64
subdivisions=4
height=64
width=64
#height=80
#width=48

channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 0
exposure = 1.5
hue=0
#max_crop=24
max_crop = 0
adam=1

learning_rate=0.001
max_batches = 5000
policy=steps
steps=600,1500,3000
scales=10,.1,.1

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

#stopbackward=1

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

###########

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=64
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=2
activation=relu

[avgpool]

[softmax]
groups=1

[cost]
type=sse

**Comment by AlexeyAB** [comment]
@agirbau Hi,
Probably it is fixed by this PR: https://github.com/AlexeyAB/darknet/pull/2542

**Comment by agirbau** [comment]
I will check it now, thank you!

**Comment by agirbau** [comment]
It seems to be working now! (image below is from training a classifier with a non-square input)

![image](https://user-images.githubusercontent.com/23292221/54831437-a0fbdf00-4cba-11e9-9044-905248937c2b.png)
